crates/gpui/src/platform/linux/wayland/clipboard.rs

```diff
@@ -0,0 +1,150 @@
+    result: Arc<Mutex<Option<Result<String>>>>,
+}
+
+impl Clipboard {
+    pub fn new() -> Self {
+        Self {
+            offer: None,
+            has_text: false,
+        }
+    }
+
+    pub fn receive_offer(self: &mut Self, offer: WlDataOffer) {
+        self.offer = Some(offer);
+        self.has_text = false;
+    }
+
+    pub fn receive_mime_type(self: &mut Self, mime: &str) {
+        self.has_text = self.has_text || mime == TEXT_MIME_TYPE;
+    }
+
+    pub fn read(self: &Self) -> Result<ClipboardRead> {
+        let Some(ref offer) = self.offer else { bail!("no clipboard data available"); };
+        Ok(ClipboardRead::new(offer.clone()))
+    }
+
+    pub fn read_sync(self: &Self) -> Option<String> {
+        let Some(ref offer) = self.offer else { return None; };
+        let read_pipe = setup_offer_read(offer.clone()).unwrap();
+
+        match read_pipe_with_timeout(read_pipe) {
+            Ok(result) => Some(result),
+            Err(_) => None,
+        }
+    }
+}
+
+impl ClipboardRead {
+    pub fn new(offer: WlDataOffer) -> Self {
+        Self {
+            offer,
+            did_start: RefCell::new(false),
+            result: Arc::new(Mutex::new(None))
+        }
+    }
+}
+
+impl Future for ClipboardRead {
+    type Output = Result<String, &'static str>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut did_start = self.did_start.borrow_mut();
+        if !*did_start {
+            *did_start = true;
+
+            let read_pipe = setup_offer_read(self.offer.clone()).unwrap();
+            let result_cell = self.result.clone();
+            let waker = cx.waker().clone();
+
+            std::thread::spawn(move || {
+                match read_pipe_with_timeout(read_pipe) {
+                    Ok(result) => {
+                        let mut result_cell = result_cell.lock();
+                        *result_cell = Some(Ok(result));
+                    }
+                    Err(e) => {
+                        log::error!("while reading clipboard: {}", e);
+                        let mut result_cell = result_cell.lock();
+                        *result_cell = Some(Err(e));
+                    }
+                };
+                waker.wake();
+            });
+        }
+
+        match &mut *self.result.lock() {
+            None => Poll::Pending,
+            Some(result) => match result {
+                Ok(result) => Poll::Ready(Ok(result.clone())),
+                Err(e) => Poll::Ready(Err("failed to read clipboard")),
+            },
+        }
+    }
+}
+
+pub(crate) fn setup_offer_read(offer: wl_data_offer::WlDataOffer) -> anyhow::Result<FileDescriptor> {
+    let pipe = Pipe::new().map_err(anyhow::Error::msg)?;
+    offer.receive(TEXT_MIME_TYPE.to_string(), unsafe { BorrowedFd::borrow_raw(pipe.write.as_raw_fd()) });
+    Ok(pipe.read)
+}
+
+pub(crate) fn read_pipe_with_timeout(mut file: FileDescriptor) -> Result<String> {
```

Comment:
This implementation comes from wezterm, just like the `filedescriptor` crate which was made for this precise use-case.